client/transaction-pool/src/revalidation.rs

```diff
@@ -0,0 +1,346 @@
+		from_queue: mpsc::UnboundedReceiver<WorkerPayload<Api>>
+	) -> Self {
+		Self {
+			api,
+			pool,
+			from_queue,
+			tick_timeout: Box::pin(interval(BACKGROUND_REVALIDATION_INTERVAL)),
+			block_ordered: Default::default(),
+			members: Default::default(),
+			best_block: Zero::zero(),
+		}
+	}
+
+	fn prepare_batch(&mut self) -> Vec<ExHash<Api>> {
+		let mut queued_exts = Vec::new();
+		let mut empty = Vec::<NumberFor<Api>>::new();
+
+		// Take maximum of count transaction by order
+		// which they got into the pool
+		for (block_number, ext_map) in self.block_ordered.iter_mut() {
+			if queued_exts.len() >= BACKGROUND_REVALIDATION_BATCH_SIZE { break; }
+
+			loop {
+				let next_key = match ext_map.iter().nth(0) {
+					Some(k) => k.clone(),
+					None => { break; }
+				};
+
+				ext_map.remove(&next_key);
+				self.members.remove(&next_key);
+
+				queued_exts.push(next_key);
+
+				if ext_map.len() == 0 { empty.push(*block_number); }
+
+				if queued_exts.len() >= BACKGROUND_REVALIDATION_BATCH_SIZE { break; }
+			}
+		}
+
+		// retain only non-empty
+		for empty_block_number in empty.into_iter() { self.block_ordered.remove(&empty_block_number); }
+
+		queued_exts
+	}
+
+	fn push(&mut self, worker_payload: WorkerPayload<Api>) {
+		// we don't add something that already scheduled for revalidation
+		let transactions = worker_payload.transactions;
+		let block_number = worker_payload.at;
+
+		for ext_hash in transactions.into_iter() {
+			// we don't add something that already scheduled for revalidation
+			if self.members.contains_key(&ext_hash) { continue; }
+
+			self.block_ordered.entry(block_number)
+				.and_modify(|value| { value.insert(ext_hash.clone()); })
+				.or_insert_with(|| {
+					let mut bt = HashSet::new();
+					bt.insert(ext_hash.clone());
+					bt
+				});
+			self.members.insert(ext_hash.clone(), block_number);
+		}
+	}
+}
+
+impl<Api: ChainApi> Future for RevalidationWorker<Api>
+where
+	Api: 'static,
+{
+	type Output = ();
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context) -> Poll<Self::Output> {
+		let this = &mut *self;
+		let mut process_some_immediately = false;
+
+		while let Poll::Ready(msg) = this.from_queue.poll_next_unpin(cx) {
+			match msg {
+				Some(worker_payload) => {
+					this.best_block = worker_payload.at;
+					this.push(worker_payload);
+					process_some_immediately = true;
+					continue;
+				},
+				// R.I.P. worker!
+				None => return Poll::Ready(()),
+			}
+		}
+
+		loop {
+			match (this.tick_timeout.poll_next_unpin(cx), process_some_immediately) {
+				(Poll::Ready(Some(_)), _) | (_, true) => {
+					process_some_immediately = false;
+
+					let next_batch = this.prepare_batch();
+					let batch_revalidate_fut =
+						batch_revalidate(this.pool.clone(), this.api.clone(), this.best_block, next_batch);
+
+					futures::pin_mut!(batch_revalidate_fut);
+					match batch_revalidate_fut.poll_unpin(cx)
+					{
+						Poll::Ready(_) => continue,
+						_ => break,
```

Comment:
That looks incorrect. We only `poll` `batch_revalidate_fut` once and drop it first time it returns `Pending`? Can't we use `async/await ` for the entire revalidation logic instead?

Reply:
This would be ok if the future of `batch_revalidate` didn't hold any state. Unfortunately that's not the case.
Being unfamiliar with the txpool, I must admit that I don't have context for the logic here.

+1 for rewriting with async/await anyway, as it's way less confusing.

Reply:
Indeed much cleaner with async/await, sorry haven't pushed myself to do it initially :)